tools/doc-generator/doc-generator.go

```diff
@@ -56,66 +95,89 @@ func parseVirtMetrics(r io.Reader, contains []string) map[string]string {
 	return metrics
 }
 
-func writeToFile(metrics map[string]string) {
-	old, err := os.Open("metrics.md")
+func writeConstMarkdown(markdown string) {
+	doc, err := os.OpenFile(docFileName, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0755)
 	checkError(err)
-	defer old.Close()
+	defer doc.Close()
 
-	new, err := os.Create("newmetrics.md")
-	checkError(err)
-	defer new.Close()
-
-	var write string
+	doc.WriteString(markdown)
+	doc.WriteString("\n")
+}
 
-	format := "###%s"
+// writeMetricsMarkdown can be used as a generic writer for metrics
+func writeMetricsMarkdown(metrics map[string]string, substring string) {
+	var markdown string
 
-	buf := bufio.NewWriter(new)
-	scan := bufio.NewScanner(old)
-	for scan.Scan() {
-		line := scan.Text()
-		if write != "" {
-			if strings.Contains(line, "HELP") {
-				if line != write {
-					buf.WriteString(write)
-					line = ""
-				}
-
-			} else {
-				buf.WriteString(write)
-				buf.WriteString("\n")
-			}
-			write = ""
+	for metric, help := range metrics {
+		if strings.Contains(metric, substring) {
+			markdown += "#### " + metric + "\n"
+			markdown += help + "\n"
+			markdown += "\n\n"
+		}
+	}
+	writeConstMarkdown(markdown)
+}
 
+// writeVMIMetricsMarkdown writes VMI metrics, while keeping related metrics together
+func writeVMIMetricsMarkdown(metrics map[string]string) {
```

Comment:
can this be generalized?

Reply:
My first try was `writeMetricsMarkdown(metrics, "kubevirt_vmi_")`, and that's when I noticed that everything is written out of order.

`writeVMIMetricsMarkdown` is what I came up with to have some low level of order with VMI metrics

Reply:
Your approach with ordering the map will solve this too :slightly_smiling_face: